File: src/main/java/myLang/expression/AdditiveExpressionContext.java
package myLang.expression;

import myLang.ContextHandler;
import myLang.block.BlockContext;
import myLang.interpreter.Function;
import myLang.interpreter.MyLangInterpreter;
import myLang.response.IntegerResponse;
import myLang.response.MyLangException;
import myLang.block.statement.FunInvocationContext;
import myLangParser.MyLangLexer;
import myLangParser.MyLangParser;
import org.antlr.v4.runtime.tree.TerminalNodeImpl;

public class AdditiveExpressionContext implements ContextHandler<MyLangParser.AdditiveExpressionContext> {

    @Override
    public IntegerResponse handler(MyLangParser.AdditiveExpressionContext ctx) throws MyLangException {
        if(ctx.NUMBER() != null){
            return new IntegerResponse(Integer.valueOf(ctx.NUMBER().getText()));
        }

        if(ctx.ID() != null){
            int result = (Integer) BlockContext.get(ctx.ID().getText()).getValue();
            return new IntegerResponse(result);
        }

        if(ctx.funInvocation() != null){
            String name = ctx.funInvocation().ID().getText();
            Function function = MyLangInterpreter.functions
                    .stream()
                    .filter(fun -> fun.getName().equals(name))
                    .findFirst()
                    .orElse(null);
            if(function == null){
                throw new MyLangException("Функция " + name + " не объявлена");
            }
            if(function.getResultType().equals("void")){
                throw new MyLangException("Функция " + name + " не возвращает результата");
            }
            return new IntegerResponse(Integer.valueOf((String) new FunInvocationContext().handler(ctx.funInvocation()).getResponse()));
        }

        if (ctx.LPAREN() != null && ctx.RPAREN() != null) {
            return handler(ctx.additiveExpression(0));
        } else {
            IntegerResponse firstResponse = handler(ctx.additiveExpression(0));
            IntegerResponse secondResponse = handler(ctx.additiveExpression(1));
            switch (((TerminalNodeImpl) ctx.getChild(1)).getSymbol().getType()) {
                case MyLangLexer.ADD:
                    return new IntegerResponse(firstResponse.getResponse() + secondResponse.getResponse());
                case MyLangLexer.SUB:
                    return new IntegerResponse(firstResponse.getResponse() - secondResponse.getResponse());
                case MyLangLexer.MUL:
                    return new IntegerResponse(firstResponse.getResponse() * secondResponse.getResponse());
                case MyLangLexer.DIV:
                    if(secondResponse.getResponse() == 0){
                        throw new MyLangException("Деление на 0");
                    }
                    return new IntegerResponse(firstResponse.getResponse() / secondResponse.getResponse());
            }
        }

        throw new MyLangException("Неподдерживаемая арифметическая операция");
    }
}

File: src/main/kotlin/myLangKotlin/expression/AdditiveExpressionContext.kt
package myLangKotlin.expression

import myLangKotlin.ContextHandler
import myLangKotlin.block.BlockContext
import myLangKotlin.block.statement.FunInvocationContext
import myLangKotlin.interpreter.MyLangInterpreter
import myLangKotlin.response.IntegerResponse
import myLangKotlin.response.MyLangException
import myLangParser.MyLangLexer
import myLangParser.MyLangParser
import org.antlr.v4.runtime.tree.TerminalNodeImpl

class AdditiveExpressionContext : ContextHandler<MyLangParser.AdditiveExpressionContext> {

    @Throws(MyLangException::class)
    override fun handler(ctx: MyLangParser.AdditiveExpressionContext): IntegerResponse {
        if (ctx.NUMBER() != null) {
            return IntegerResponse(Integer.valueOf(ctx.NUMBER().text))
        }

        if (ctx.ID() != null) {
            val result = BlockContext[ctx.ID().text].value as Int
            return IntegerResponse(result)
        }

        if (ctx.funInvocation() != null) {
            val name = ctx.funInvocation().ID().text
            val function = MyLangInterpreter.functions
                    .stream()
                    .filter { `fun` -> `fun`.name == name }
                    .findFirst()
                    .orElse(null) ?: throw MyLangException("Функция $name не объявлена")
            if (function.resultType == "void") {
                throw MyLangException("Функция $name не возвращает результата")
            }
            return IntegerResponse(Integer.valueOf(FunInvocationContext().handler(ctx.funInvocation()).response as String))
        }

        if (ctx.LPAREN() != null && ctx.RPAREN() != null) {
            return handler(ctx.additiveExpression(0))
        } else {
            val firstResponse = handler(ctx.additiveExpression(0))
            val secondResponse = handler(ctx.additiveExpression(1))
            when ((ctx.getChild(1) as TerminalNodeImpl).getSymbol().type) {
                MyLangLexer.ADD -> return IntegerResponse(firstResponse.response + secondResponse.response)
                MyLangLexer.SUB -> return IntegerResponse(firstResponse.response - secondResponse.response)
                MyLangLexer.MUL -> return IntegerResponse(firstResponse.response * secondResponse.response)
                MyLangLexer.DIV -> {
                    if (secondResponse.response == 0) {
                        throw MyLangException("Деление на 0")
                    }
                    return IntegerResponse(firstResponse.response / secondResponse.response)
                }
            }
        }

        throw MyLangException("Неподдерживаемая арифметическая операция")
    }
}
